src/mame/dataeast/dec8.h

```diff
@@ -17,216 +17,299 @@
-	optional_shared_ptr<uint8_t> m_bg_data;
+	required_shared_ptr<u8> m_videoram;
+	optional_shared_ptr<u8> m_bg_ram;
 
-	uint8_t *  m_pf1_data = nullptr;
-	uint8_t *  m_row = nullptr;
-	std::unique_ptr<uint16_t[]>   m_buffered_spriteram16; // for the mxc06 sprite chip emulation (oscar, cobra)
+	std::unique_ptr<u16[]>   m_buffered_spriteram16; // for the mxc06 sprite chip emulation (oscar, cobra)
 
 	/* video-related */
 	tilemap_t  *m_bg_tilemap = nullptr;
-	tilemap_t  *m_pf1_tilemap = nullptr;
 	tilemap_t  *m_fix_tilemap = nullptr;
-	//int      m_scroll1[4]{};
-	int      m_scroll2[4]{};
-	int      m_bg_control[0x20]{};
-	int      m_pf1_control[0x20]{};
+	int      m_scroll[4]{};
 	int      m_game_uses_priority = 0;
 
 	/* misc */
-	bool     m_secclr = false;
-	bool     m_nmi_enable = false;
-	int      m_coinage_id = 0;
-	int      m_coin1 = 0;
-	int      m_coin2 = 0;
-	int      m_need1 = 0;
-	int      m_need2 = 0;
-	int      m_cred1 = 0;
-	int      m_cred2 = 0;
-	int      m_credits = 0;
-	int      m_latch = 0;
 	bool     m_coin_state = false;
-	int      m_snd = 0;
 
-	emu_timer *m_i8751_timer = nullptr;
 	emu_timer *m_m6502_timer = nullptr;
 
-	void i8751_reset_w(uint8_t data);
-	uint8_t gondo_player_1_r(offs_t offset);
-	uint8_t gondo_player_2_r(offs_t offset);
-	void dec8_bank_w(uint8_t data);
-	void ghostb_bank_w(uint8_t data);
-	void sub_firq_off_w(uint8_t data);
-	void gondo_mcu_to_main_w(uint8_t data);
-	void shackled_mcu_to_main_w(uint8_t data);
-	void srdarwin_mcu_to_main_w(uint8_t data);
-	void srdarwin_videoram_w(offs_t offset, uint8_t data);
-	void srdarwin_control_w(offs_t offset, uint8_t data);
-	void lastmisn_control_w(uint8_t data);
-	void shackled_control_w(uint8_t data);
-	void lastmisn_scrollx_w(uint8_t data);
-	void lastmisn_scrolly_w(uint8_t data);
-	void gondo_scroll_w(offs_t offset, uint8_t data);
-	TILE_GET_INFO_MEMBER(get_cobracom_fix_tile_info);
-	TILE_GET_INFO_MEMBER(get_ghostb_fix_tile_info);
-	TILE_GET_INFO_MEMBER(get_oscar_fix_tile_info);
+	void allocate_buffered_spriteram16();
+	void dec8_s_map(address_map &map);
+	void oscar_s_map(address_map &map);
+
+};
+
+// with I8751 MCU
+class mcu_state : public base_state
+{
+public:
+	mcu_state(const machine_config &mconfig, device_type type, const char *tag) :
+		base_state(mconfig, type, tag),
+		m_mcu(*this, "mcu"),
+		m_coin_port(*this, "I8751")
+	{
+	}
+protected:
+	virtual void machine_start() override;
+	virtual void machine_reset() override;
+
+	TIMER_CALLBACK_MEMBER(mcu_irq_clear);
+
+	required_device<i8751_device> m_mcu;
+
+	/* ports */
+	required_ioport m_coin_port;
+
+	u8 i8751_h_r();
+	u8 i8751_l_r();
+	void i8751_w(offs_t offset, u8 data);
+
+	u8 i8751_port0_r();
+	void i8751_port0_w(u8 data);
+	u8 i8751_port1_r();
+	void i8751_port1_w(u8 data);
+
+	void i8751_reset_w(u8 data);
+
+	// MCU communication
+	u8  m_i8751_p2 = 0;
+	int m_i8751_port0 = 0;
+	int m_i8751_port1 = 0;
+	int m_i8751_return = 0;
+	int m_i8751_value = 0;
+
+	emu_timer *m_i8751_timer = nullptr;
```

Comment:
Please keep data members grouped – you’ve got `m_mcu` and `m_coin_port` separate from the rest of the data members.